### PR TITLE
Liberty

### PR DESCRIPTION
**Pull request by foolswood**
This broadly outlines all the things I can remember liberty being for.

### CODE REVIEW

File: src/Clapi/Valuespace.hs [hunk 1]
```diff
@@ -477,12 +477,31 @@ vsClientValidate vs = vsDiff ovs <$> doValidate mempty vs
...
+  where
+    libertyOf name (TupleDef {}) = fail "Tuples have no children"
+    libertyOf name (StructDef _ names _ clibs) = note "Not a child of parent struct" $ lookup name $ zip names clibs
+    libertyOf name (ArrayDef _ _ clib) = return clib
+
+checkLibertiesPermit :: Valuespace ClientUnvalidated -> MonadErrorMap (Valuespace ClientUnvalidated)
+checkLibertiesPermit vs = (errs, vs)
+  where
+    taintedPaths = Map.keys $ view (unvalidated . uvtt) vs
+    liberties = zip taintedPaths (map (flip getLiberty vs) taintedPaths :: [Maybe Liberty])
```
Comment (ch3pjw) [inline src/Clapi/Valuespace.hs]: Why define `getLiberty` the way around you do if you just need to flip it here?

File: src/Clapi/Valuespace.hs [hunk 2]
```diff
@@ -502,6 +504,36 @@ checkLibertiesPermit vs = (errs, vs)
...
+  where
+    checkPath (np, (node, def)) = (validateClientNodeChildren getOldNode getNewNode np node def, ())
+    getOldNode np = Map.lookup np $ vsGetTree vs
+    getNewNode np = Map.lookup np $ vsGetTree $ origVs $ view unvalidated vs
+
+validateClientNodeChildren :: (NodePath -> Maybe Node) -> (NodePath -> Maybe Node) -> NodePath -> Node -> Definition -> ErrorMap
+validateClientNodeChildren getOldNode getNewNode np node def = Map.fromList $ case def of
+    (TupleDef {}) -> if null childKeys then mempty else map (\p -> (p, "Tuples have no children!")) childPaths
+    (StructDef _doc childNames _childTypes childLiberties) -> let
+        nameLibMap = Map.fromList $ zip  childNames childLiberties
```
Comment (ch3pjw) [inline src/Clapi/Valuespace.hs]: Bonus space

File: src/Clapi/Valuespace.hs [hunk 3]
```diff
@@ -502,6 +504,36 @@ checkLibertiesPermit vs = (errs, vs)
...
+        getChildErr name = let sp = childPath name in
+            fmap (\e -> (sp, e)) $ case Map.lookup name nameLibMap of
```
Comment (ch3pjw) [inline src/Clapi/Valuespace.hs]: Could `fmap ((,) sp)` (partially apply the pair constructor)

File: src/Clapi/Valuespace.hs (hunk 1 above)
Comment (foolswood) [reply src/Clapi/Valuespace.hs]: Was trying to be consistent about taking the vs as the last arg, will flip.

**Comment by foolswood** [comment]
Haddock internal errors, great :(

**Pull request merged**